### ISSUE TITLE
Workflow structure

### ISSUE THREAD

**Issue by andersgs** [jts/ncov-tools#47]
Hi Jared.

Excellent work. Here are some suggestions that may help in the long term maintainability and portability of the project.

It would be helpful if the workflow were restructured to what is proposed as the recommended structure for Snakemake workflows: https://github.com/snakemake/snakemake/blob/fdc26e642f08fb4ca3f66fe8e310e8eda8a954a6/docs/snakefiles/deployment.rst

This would carry the advantages of:
* Eliminating the need for all the calls to `srcdir` when sourcing the paths to the various scripts. 
* Eliminating the need to explicitly refer to the config file in the Snakefile
* Removing all the function definitions to a `rules/common.smk` and just called into the main Snakefile with `include: rules/common.smk, making thins more readable and easier to maintain
* The different targets could be refactored to `rules/target1.smk` and run if requested with (**pseudocode**) `if target include target: include: rules/target1.smk`, which would be easier to maintain, and to see changes over time in your commit history - not to mention merge any PR requests
* you can have `envs' for each rule/target - easier to maintain (lets say one target needs bedtools <1.9, while another uses a new feature of bedtools? - contrived, but you get the picture)
* to avoid having to pass in variables to the python scripts in the params block, the scripts could be re-written to take advantage of the `snakemake` object that is made available when using the `scripts` directive instead of the `shell` directive. With the new structure, you would call a script with:

```
script:
  script/myscript.py
```

And reference access all the inputs/outputs/config/wildcards, etc with the `snakemake` object: https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#external-scripts

If you still wished to have the scripts as standalone scripts that take arguments, you could so something like this:

```
def get_args(snakemake=None):
  if snakemake:
    from collections import namedtuple
    Args = namedtupble('Args', ['read1', 'read2', 'outfile', 'threads'])
    args = Args(read1 = snakemake.input['read1'], read2 = snakemake.input['read2'], outfile=snakemake.output[0], threads=snakemake.threads)
    return args
  args = argparse.ArgumentParser(...)
  return args

try:
  args = get_args(snakemake)
except:
  args = get_args()
```

Finally, the `parser` package should be spun out into a separate repository and made pip installable.

If these are things you would be at all interested in, I can help with the refactor. But, did not want to start anything without discussing first.

Cheers. Anders.

**Comment by jts** [comment]
Thanks for the comments Anders! @rdeborja and I will discuss and get back to you

**Comment by jts** [comment]
Hi @andersgs,

I have a question about this proposal. Typically we store the Snakefile in a central repo, then make config.yaml in the directory containing the run we want to process. For example:

```
sequencing_run_abcd
   data/
      sampleA.bam
      sampleA.variants.tsv
      [...]
   config.yaml
sequencing_run_efgh
   [...]
```

Then we run snakemake with `-s <path>/qc/Snakefile` within `sequencing_run_abcd` to populate the results directories:

```
sequencing_run_abcd
   data/
      sampleA.bam
      sampleA.variants.tsv
      [...]
   qc_sequencing/
      [...]
   qc_analysis/
      [...]
   qc_reports/
      [...]
   config.yaml
sequencing_run_efgh
   [...]
```

If I understand correctly in the recommended deployment structure it is intended that each run directory should get its own clone of the complete workflow (Snakefiles, scripts, config, envs). Is it possible to adhere to the proposed deployment structure but separate the code and config? I'd like to continue to have the code in a central location and keep the config elsewhere. I'm happy to hear arguments in favour of keeping it all together though.

Jared

**Comment by andersgs** [comment]
@jts that is a fair point. the argument against that workflow is that you have no way of knowing what version of the workflow was run in each run. so, you would have a hard time reproducing the results if you needed to for a specific run in the future - which is a typical requirement for public health bioinformatic workflows. I am assuming that you run the latest version of the workflow for each new run, but don't necessary go back to re-process past runs with newer versions of the workflow.

i typically have a workflow and a runner script, and we deploy them via docker, and keep track of the docker/singularity container that was run for each set of results.

to ensure separation of concerns, you could have two repos. One dedicated to the workflow, that makes tagged releases for versions that are production ready (where you can have tests for individual components of the workflow). A second with a runner script that will clone (if needed) a particular version of the workflow (based on the tag --- the cloning can be to a central location or to the run folder), will take arguments to configure the `config.yml` file, and you could have a `--conda-prefix` to make sure environments are not installed again with every run. The runner script would be in charge of running the workflow, and outputting a provenance yaml with some basic information of the run --- this could include the parameters that would typically be in the config.yml file. The runner script should be pip installable, and that is what a typical user would install.

The runner script can be generic, and allow one to point to different workflow repos. So, it doesn't have to be tied down to the workflow. The runner script could include options to run the workflow on a worker node on an HPC cluster using `dask` and `dask-jobqueue`, for instance. Something I have been experimenting with.

The runner script can take care of running the snakefile by importing `snakemake` and configuring the `snakemake.snakemake` object directly rather than using a `subprocess`. In this scenario, you would need a config dictionary and not a config.yml file, so that is why I suggest adding the parameters for the specific configuration to the provenance yaml file above.

happy to discuss further.

**Comment by jts** [comment]
Thanks - a few comments:

> I am assuming that you run the latest version of the workflow for each new run, but don't necessary go back to re-process past runs with newer versions of the workflow.

Right now we always re-process all runs when we update `ncov-tools` but in the future as we release data publicly, or the results are included in publications, we'd freeze the results and not re-process. I agree with your points about provenance - usually (for other projects) I write the commit ID or version tag somewhere in the results to track this. Bundling the code as in the full deployment model is an interesting approach - I'll think about this a bit more.

>  One dedicated to the workflow, that makes tagged releases for versions that are production ready (where you can have tests for individual components of the workflow). A second with a runner script that will clone (if needed) a particular version of the workflow (based on the tag --- the cloning can be to a central location or to the run folder), will take arguments to configure the config.yml file

This is what we currently do internally, although the runner script is not public. It generates the config.yaml and runs the appropriate targets with the snakemake.

**Comment by jts** [comment]
Hi @andersgs,

I have started to try your suggestions in the `refactor` branch (referenced above). I have adopted the `workflow` structure and created files in `rules` for the main parts of the code (I still have to clean it up a lot, this is just a first cut). I have placed the external python/R code in `scripts`.

I was unable to remove the `srcdir` calls in this pass. It appears we cannot pass arbitrary arguments when using `script:` in the rule so the input/output/wildcards need to be read from the snakemake object. While some of our scripts exist only to be run within the Snakefile, many others also need a command line interface. I haven't decided yet if I want to keep the ugly `srcdir` calls and continue to use `shell`, or rewrite all the scripts to use the snakemake object and your wrapper (`def get_args(snakemake=None)`). 

Thanks again for the suggestions!
